controllers/image_controller.go

```diff
@@ -0,0 +1,255 @@
+	"time"
+)
+
+// ImageReconciler reconciles a Image object
+type ImageReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+type ImageSpecJobAction int
+
+const (
+	unknown ImageSpecJobAction = iota
+	create
+	cancel
+	rebuild
+	update
+	skip
+)
+
+func makeImageControllerAction(r *ImageReconciler, ctx context.Context, image *v1alpha1.Image) (ImageSpecJobAction, *v1alpha1.ImageSpecJob) {
+	if image != nil && image.Spec.ImageSpec.BaseImage != "" {
+		// Get ImageSpecJob by Image
+		imageSpecJob := &v1alpha1.ImageSpecJob{}
+		imageSpecJobName := getImageSpecJobName(image)
+		err := r.Get(ctx, client.ObjectKey{Namespace: image.Namespace, Name: imageSpecJobName}, imageSpecJob)
+
+		// Create if no ImageSpecJob found
+		if apierrors.IsNotFound(err) {
+			if image.Spec.ImageSpec.Cancel == false {
+				return create, nil
+			}
+			return skip, nil
+		}
+
+		// Cancel if cancel flag is marked in Image
+		if image.Spec.ImageSpec.Cancel == true && image.Status.JobCondiction.Phase != CustomImageJobStatusCancelled {
+			return cancel, imageSpecJob
+		}
+
+		// Rebuild if image phase was succeed or failed and image.updateTime after imageSpecJob.updateTime
+		imageUpdateTime := image.Spec.ImageSpec.UpdateTime
+		imageSpecJobUpdateTime := imageSpecJob.Spec.UpdateTime
+		if imageUpdateTime != nil && imageSpecJobUpdateTime != nil &&
+			imageUpdateTime.After(imageSpecJob.Spec.UpdateTime.Time) {
+			return rebuild, imageSpecJob
+		}
+
+		// Update if imageSpecJob exist
+		return update, imageSpecJob
+	}
+	return unknown, nil
+}
+
+func isImageCustomBuild(image *v1alpha1.Image) bool {
+	if image == nil || image.Spec.ImageSpec.BaseImage == "" {
+		return false
+	}
+	return true
+}
+
+func getImageSpecJobName(image *v1alpha1.Image) string {
+	if image == nil {
+		return ""
+	}
+	if image.Spec.ImageSpec.BaseImage == "" {
+		return ""
+	}
+	return image.Name
+}
+
+func createImageSpecJob(r *ImageReconciler, ctx context.Context, image *v1alpha1.Image) error {
+	if !isImageCustomBuild(image) {
+		return fmt.Errorf("creiateImageSpecJob do not provide correct image")
+	}
+
+	pushSecretName := viper.GetString("customImage.pushSecretName")
+	repoPrefix := viper.GetString("customImage.pushRepoPrefix")
+
+	hash := computeHash(image.Spec.ImageSpec.BaseImage, image.Spec.ImageSpec.Packages)
+
+	updateTime := image.Spec.ImageSpec.UpdateTime
+	if updateTime == nil {
+		updateTime = &metav1.Time{Time: time.Now()}
+	}
+
+	imageSpecJob := &v1alpha1.ImageSpecJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        getImageSpecJobName(image),
+			Namespace:   image.Namespace,
+			Annotations: map[string]string{"image.primehub.io/hash": hash},
+			Labels: map[string]string{
+				"image.primehub.io/name": image.ObjectMeta.Name,
+				"app":                    "primehub-image",
+			},
+		},
+		Spec: v1alpha1.ImageSpecJobSpec{
+			BaseImage:   image.Spec.ImageSpec.BaseImage,
+			PullSecret:  image.Spec.PullSecret,
```

Comment:
Copy wrong field. Should by `image.ImageSpec.PullSecret`

Reply:
Fixed